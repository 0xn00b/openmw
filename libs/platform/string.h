// Wrapper for string.h on Mac
#ifndef _STRING_WRAPPER_H
#define _STRING_WRAPPER_H

#include <string.h>
#ifdef __APPLE__
// need our own implementation of strnlen
static size_t strnlen(const char *s, size_t n)
{
  const char *p = (const char *)memchr(s, 0, n);
  return(p ? p-s : n);
}
#endif
#endif
