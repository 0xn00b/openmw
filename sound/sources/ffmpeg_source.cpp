#include "ffmpeg_source.hpp"

#include "../../tools/str_exception.hpp"

using namespace Mangle::Sound;

// Static output buffer. Not thread safe, but supports multiple
// streams operated from the same thread.
static uint8_t outBuf[AVCODEC_MAX_AUDIO_FRAME_SIZE];

static void fail(const std::string &msg)
{ throw str_exception("FFMpeg exception: " + msg); }

// --- Loader ---

static bool init = false;

FFMpegLoader::FFMpegLoader(bool setup)
{
  if(setup && !init)
    {
      av_register_all();
      av_log_set_level(AV_LOG_ERROR);
      init = true;
    }
}

// --- Source ---

FFMpegSource::FFMpegSource(const std::string &file)
{
  std::string msg;
  AVCodec *codec;

  if(av_open_input_file(&FmtCtx, file.c_str(), NULL, 0, NULL) != 0)
    fail("Error loading audio file " + file);
  
  if(av_find_stream_info(FmtCtx) < 0)
    {
      msg = "Error in file stream " + file;
      goto err;
    }

  // Pick the first audio stream, if any
  for(StreamNum = 0; StreamNum < FmtCtx->nb_streams; StreamNum++)
    {
      // Pick the first audio stream
      if(FmtCtx->streams[StreamNum]->codec->codec_type == CODEC_TYPE_AUDIO)
        break;
    }

  if(StreamNum == FmtCtx->nb_streams)
    fail("File '" + file + "' didn't contain any audio streams");

  // Open the decoder
  CodecCtx = FmtCtx->streams[StreamNum]->codec;
  codec = avcodec_find_decoder(CodecCtx->codec_id);

  if(!codec || avcodec_open(CodecCtx, codec) < 0)
    {
      msg = "Error loading '" + file + "': ";
      if(codec)
        msg += "coded error";
      else
        msg += "no codec found";
      goto err;
    }

  // No errors, we're done
  return;

  // Handle errors
 err:
  av_close_input_file(FmtCtx);
  fail(msg);
}

FFMpegSource::~FFMpegSource()
{
  avcodec_close(CodecCtx);
  av_close_input_file(FmtCtx);
}

void FFMpegSource::getInfo(int32_t *rate, int32_t *channels, int32_t *bits)
{
  if(rate) *rate = CodecCtx->sample_rate;
  if(channels) *channels = CodecCtx->channels;
  if(bits) *bits = 16;
}

size_t FFMpegSource::read(void *data, size_t length)
{
  if(isEof) return 0;

  size_t left = length;
  uint8_t *outPtr = (uint8_t*)data;

  // First, copy over any stored data we might be sitting on
  {
    size_t s = storage.size();
    size_t copy = s;
    if(s)
      {
        // Make sure there's room
        if(copy > left)
          copy = left;

        // Copy
        memcpy(outPtr, &storage[0], copy);
        outPtr += copy;
        left -= copy;

        // Is there anything left in the storage?
        assert(s>= copy);
        s -= copy;
        if(s)
          {
            assert(left == 0);

            // Move it to the start and resize
            memmove(&storage[0], &storage[copy], s);
            storage.resize(s);
          }
      }
  }

  // Next, get more input data from stream, and decode it
  while(left)
    {
      AVPacket packet;

      // Get the next packet, if any
      if(av_read_frame(FmtCtx, &packet) < 0)
        break;

      // We only allow one stream per file at the moment
      assert((int)StreamNum == packet.stream_index);

      // Decode the packet
      int len = AVCODEC_MAX_AUDIO_FRAME_SIZE;
      int tmp = avcodec_decode_audio2(CodecCtx, (int16_t*)outBuf,
                                      &len, packet.data, packet.size);
      assert(tmp < 0 || tmp == packet.size);

      // We don't need the input packet any longer
      av_free_packet(&packet);

      if(tmp < 0)
        fail("Error decoding audio stream");

      // Copy whatever data we got, and advance the pointer
      if(len > 0)
        {
          // copy = how many bytes do we copy now
          size_t copy = len;
          if(copy > left)
            copy = left;

          // len = how many bytes are left uncopied
          len -= copy;

          // copy data
          memcpy(outPtr, outBuf, copy);

          // left = how much space is left in the caller output
          // buffer. This loop repeats as long left is > 0
          left -= copy;
          outPtr += copy;
          assert(left >= 0);

          if(len > 0)
            {
              // There were uncopied bytes. Store them for later.
              assert(left == 0);
              storage.resize(len);
              memcpy(&storage[0], outBuf, len);
            }
        }
    }

  // End of loop. Return the number of bytes copied.
  assert(left <= length);

  // If we're returning less than asked for, then we're done
  if(left > 0)
    isEof = true;

  return length - left;
}
