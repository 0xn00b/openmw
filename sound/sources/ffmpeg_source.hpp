#ifndef MANGLE_SOUND_FFMPEG_H
#define MANGLE_SOUND_FFMPEG_H

#include "../source.hpp"
#include <vector>
#include <assert.h>

extern "C"
{
#include <avcodec.h>
#include <avformat.h>
}

namespace Mangle {
namespace Sound {

class FFMpegSource : public SampleSource
{
  AVFormatContext *FmtCtx;
  AVCodecContext *CodecCtx;
  unsigned int StreamNum;

  std::vector<uint8_t> storage;

 public:
  /// Decode the given sound file
  FFMpegSource(const std::string &file);

  /// Decode the given sound stream (not supported by FFmpeg)
  FFMpegSource(Mangle::Stream::StreamPtr src) { assert(0); }

  ~FFMpegSource();

  // Overrides
  void getInfo(int32_t *rate, int32_t *channels, int32_t *bits);
  size_t read(void *data, size_t length);
};

#include "loadertemplate.hpp"

/// A factory that loads FFMpegSources from file
class FFMpegLoader : public SSL_Template<FFMpegSource,false,true>
{
 public:

  /// Sets up the libavcodec library. If you want to do your own
  /// setup, send a setup=false parameter.
  FFMpegLoader(bool setup=true);
};

}} // namespaces
#endif
