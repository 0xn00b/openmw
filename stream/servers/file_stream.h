#ifndef MANGLE_STREAM_FILESERVER_H
#define MANGLE_STREAM_FILESERVER_H

#include "std_stream.h"
#include <fstream>

namespace Mangle {
namespace Stream {

/** Very simple file input stream, based on std::ifstream
 */
class FileStream : public StdStream
{
  std::ifstream file;

 public:
  FileStream(const std::string &name)
    : StdStream(&file)
    {
      file.open(name, ios::binary);
    }
  ~FileStream() { file.close(); }
};

typedef boost::shared_ptr<FileStream> FileStreamPtr;

}} // namespaces
#endif
