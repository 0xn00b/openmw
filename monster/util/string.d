/*
  Monster - an advanced game scripting language
  Copyright (C) 2004, 2007, 2008  Nicolay Korslund
  Email: <korslund@gmail.com>
  WWW: http://monster.snaptoad.com/

  This file (string.d) is part of the Monster script language
  package.

  Monster is distributed as free software: you can redistribute it
  and/or modify it under the terms of the GNU General Public License
  version 3, as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  version 3 along with this program. If not, see
  http://www.gnu.org/licenses/ .

 */

module monster.util.string;

import std.utf;
import std.string;

// These functions check whether a string begins or ends with a
// certain substring.
bool begins(char[] str, char[] start)
{
  if(str.length < start.length ||
     str[0..start.length] != start) return false;
  return true;
}

bool ends(char[] str, char[] end)
{
  if(str.length < end.length ||
     str[$-end.length..$] != end) return false;
  return true;
}

// Case insensitive versions of begins and ends
bool iBegins(char[] str, char[] start)
{
  if(str.length < start.length ||
     icmp(str[0..start.length], start) != 0) return false;
  return true;
}

bool iEnds(char[] str, char[] end)
{
  if(str.length < end.length ||
     icmp(str[$-end.length..$], end) != 0) return false;
  return true;
}

char[] nextWord(ref char[] str, char delim = ' ')
{
  int i = find(str, delim);
  char[] result;

  // No 'delim' found, return the entire string and set remainder to
  // null.
  if(i == -1)
    {
      result = str;
      str = null;
      return result;
    }

  // A separator was found. Return everything upto 'delim' (index i),
  // put the remainder of the string (not including the char at [i])
  // in str.
  result = str[0..i];
  str = str[i+1..$];
  return result;
}

// Strip trailing zeros
char[] stripz(char [] s)
{
  foreach(int i, char c; s)
    if( c == 0 )
      return s[0..i];

  return s;
}

// Convert a long integer into a string using nice comma
// formatting. delim is the delimiter character, size is the number of
// digits in each group. See the unittest for examples.
char[] comma(long i, char delim=',', int size = 3)
{
  char[] str = toString(i);
  char[] res;

  if(i<0) str=str[1..$];

  str.reverse;
  foreach(int j, char c; str)
    { 
      if(j!=0 && j%size == 0)
        res = delim ~ res;
      res = c ~ res;
    }

  if(i<0) res = "-" ~ res;

  return res;
}
