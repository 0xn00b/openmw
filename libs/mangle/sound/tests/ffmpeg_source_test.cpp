#include <iostream>

#include "../../stream/servers/file_stream.hpp"
#include "../sources/ffmpeg_source.hpp"

#include <assert.h>
#include <string.h>

using namespace std;
using namespace Mangle::Stream;
using namespace Mangle::Sound;

// Contents and size of cow.raw
void *orig;
size_t orig_size;

void run(SampleSourcePtr &src)
{
  int rate, channels, bits;
  src->getInfo(&rate, &channels, &bits);
  cout << "rate=" << rate << "\nchannels=" << channels
       << "\nbits=" << bits << endl;

  cout << "Reading entire buffer into memory\n";
  void *buf = malloc(orig_size);
  size_t ss = src->read(buf, orig_size);
  cout << "Actually read: " << ss << endl;
  assert(ss == orig_size);

  cout << "Comparing...\n";
  if(memcmp(buf, orig, ss) != 0)
    {
      cout << "Oops!\n";
      assert(0);
    }

  cout << "Done\n";
}

int main()
{
  {
    cout << "Reading cow.raw first\n";
    FileStream tmp("cow.raw");
    orig_size = tmp.size();
    cout << "Size: " << orig_size << endl;
    orig = malloc(orig_size);
    tmp.read(orig, orig_size);
    cout << "Done\n";
  }

  // Initializes the library, not used for anything else.
  FFMpegLoader fm;

  {
    cout << "\nLoading cow.wav by filename:\n";
    SampleSourcePtr cow_file( new FFMpegSource("cow.wav") );
    run(cow_file);
  }

  return 0;
}
