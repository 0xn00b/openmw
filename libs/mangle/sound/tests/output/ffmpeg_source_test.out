Reading cow.raw first
Size: 37502
Done

Loading cow.wav by filename:
rate=11025
channels=1
bits=16
Reading entire buffer into memory
Actually read: 37502
Comparing...
Done
